File: modules/06b-langgraph/py/01_state_reducers.py
"""
Task 1 — State, channels & reducers  🟡

What this teaches:
  - State is a typed dict; each field is a CHANNEL with a REDUCER.
  - A node returns ONLY the keys it changes. LangGraph merges each key through
    its channel's reducer:
      * default reducer (no annotation) = last-write-wins (overwrite)
      * operator.add on an int          = accumulate (sum)
      * add_messages                    = append + dedupe-by-id, coerce -> Messages
  - "Designing the state" is mostly "pick the right reducer per channel."

This file RUNS as-is (no model/API key needed). Your job is to PREDICT the output
first, then change a reducer and predict again — the exercises are in main().

How to run:
  uv run python modules/06b-langgraph/py/01_state_reducers.py
"""

from __future__ import annotations

import operator
from typing import Annotated, TypedDict

from langchain_core.messages import AIMessage, AnyMessage, HumanMessage
from langgraph.graph import END, START, StateGraph, add_messages

# ---------------------------------------------------------------------------
# State — three channels, three DIFFERENT reducers
# ---------------------------------------------------------------------------


class DemoState(TypedDict):
    messages: Annotated[list[AnyMessage], add_messages]  # append + dedupe
    step_count: Annotated[int, operator.add]  # SUM each update
    last_tool: str  # no reducer = overwrite


# ---------------------------------------------------------------------------
# Nodes — each returns a PARTIAL update to all three channels
# ---------------------------------------------------------------------------


def node_a(state: DemoState) -> dict:
    return {
        "messages": [AIMessage(content="node A ran")],
        "step_count": 1,
        "last_tool": "search",
    }


def node_b(state: DemoState) -> dict:
    return {
        "messages": [AIMessage(content="node B ran")],
        "step_count": 1,
        "last_tool": "calculator",
    }


# ---------------------------------------------------------------------------
# Build — START -> a -> b -> END
# ---------------------------------------------------------------------------

graph = StateGraph(DemoState)
graph.add_node("a", node_a)
graph.add_node("b", node_b)
graph.add_edge(START, "a")
graph.add_edge("a", "b")
graph.add_edge("b", END)
app = graph.compile()


def main() -> None:
    initial: DemoState = {
        "messages": [HumanMessage(content="go")],
        "step_count": 0,
        "last_tool": "",
    }

    # TODO 1: PREDICT each line before you read the output.
    #   After both nodes run, what are messages / step_count / last_tool?
    #   (Answer: 3 messages, step_count 2, last_tool "calculator" — verify below.)
    for snapshot in app.stream(initial, stream_mode="values"):
        print(
            f"messages={len(snapshot['messages'])}  "
            f"step_count={snapshot['step_count']}  "
            f"last_tool={snapshot['last_tool']!r}"
        )

    # TODO 2: Change step_count's reducer (line ~36) from `operator.add` to a bare
    #         `int` (overwrite). PREDICT the new step_count, then re-run.
    #         (It becomes 1, not 2 — the last write wins instead of summing.)
    #
    # TODO 3: Do the same thought experiment for `messages`: drop `add_messages`
    #         and predict what `len(messages)` becomes. (Overwrite -> 1.)


if __name__ == "__main__":
    main()

File: modules/06b-langgraph/ts/01-state-reducers.ts
/**
 * Task 1 — State, channels & reducers  🟡
 *
 * What this teaches:
 *   - State is built with Annotation.Root; each field is a CHANNEL with a REDUCER.
 *   - A node returns ONLY the keys it changes; LangGraph merges each key through
 *     its channel's reducer:
 *       * no reducer            = last-write-wins (overwrite)
 *       * (a, b) => a.concat(b) = append (lists, like messages)
 *       * (a, b) => a + b       = accumulate (sum a counter)
 *   - "Designing the state" is mostly "pick the right reducer per channel."
 *
 * How to run:
 *   pnpm tsx modules/06b-langgraph/ts/01-state-reducers.ts
 *
 * No model/API key needed — pure graph mechanics.
 */

import { StateGraph, START, END, Annotation } from "@langchain/langgraph";
import { AIMessage, HumanMessage, BaseMessage } from "@langchain/core/messages";

// ---------------------------------------------------------------------------
// State
// ---------------------------------------------------------------------------

// TODO 1: give each channel the right reducer.
const DemoState = Annotation.Root({
  messages: Annotation<BaseMessage[]>({
    reducer: (a, b) => a.concat(b), // append
    default: () => [],
  }),
  stepCount: Annotation<number>({
    // TODO: make this SUM updates instead of overwriting.
    reducer: (_a, b) => b, // <-- overwrite for now; change to (a, b) => a + b
    default: () => 0,
  }),
  lastTool: Annotation<string>(), // no reducer => overwrite (last write wins)
});

type DemoStateT = typeof DemoState.State;

// ---------------------------------------------------------------------------
// Nodes — each returns a PARTIAL update to all three channels
// ---------------------------------------------------------------------------

const nodeA = (_state: DemoStateT) => ({
  messages: [new AIMessage("node A ran")],
  stepCount: 1,
  lastTool: "search",
});

const nodeB = (_state: DemoStateT) => ({
  messages: [new AIMessage("node B ran")],
  stepCount: 1,
  lastTool: "calculator",
});

// ---------------------------------------------------------------------------
// Build
// ---------------------------------------------------------------------------

// TODO 2: wire START -> a -> b -> END and compile.
const app = new StateGraph(DemoState)
  .addNode("a", nodeA)
  .addNode("b", nodeB)
  .addEdge(START, "a")
  // .addEdge("a", "b")
  // .addEdge("b", END)
  .compile();

async function main() {
  const initial = { messages: [new HumanMessage("go")], stepCount: 0, lastTool: "" };

  // TODO 3: stream with streamMode "values" and print state after each step.
  //   PREDICT first (with the SUM reducer): messages=3, stepCount=2, lastTool="calculator".
  for await (const snap of await app.stream(initial, { streamMode: "values" })) {
    console.log(
      `messages=${snap.messages.length}  stepCount=${snap.stepCount}  lastTool=${JSON.stringify(snap.lastTool)}`,
    );
  }

  // TODO 4: flip stepCount back to overwrite; predict (-> 1), then re-run.
}

main().catch(console.error);
